### ISSUE TITLE
This doesn't scale.

### ISSUE THREAD

**Issue by edgework** [ProjectOpenSea/stream-js#119]
I guess this thing is cool for creators who run a single instance of their server/app on a Heroku instance but for larger organizations (like the one I'm working for) that run multiple instances of the same server in scaling groups in various datacenters the stream-api is an unworkable nightmare.   How do we reconcile the message when it's received 100 or 1000 times across 3 data centers?  What if the socket message is sent during downtime (scheduled or not?) They're gone.  Already, I'm losing 1/3 of messages with a single connection that stays up while I'm doing things in the store that should send an event.  (Well actually after you got rid of the API key in dev its totally broken)

We can't run a single instance of some server to just listen for this because it just doesn't work that way when you're not a one man shop on Heroku.  I'm sure websockets are fun and cool and all but they are a very bad solution for what you are trying to accomplish.  Normally,  you would have clients register a callback URL (Webhooks) that your severs call until a 200 response is received. That's the way "web2" does it and it works awesome.

Please rethink this whole thing and implement event URL registration.

**Comment by risingsun007** [comment]
Why are websockets a bad idea?

**Comment by rick-rheo** [comment]
> Why are websockets a bad idea?


As I said, this works fine if you have one server running in Heroku or something.  When you're a large company running many instances of the server in a scaling group you've got many websocket connections all getting the same message. How do you reconcile that? Do you update the local database with the same data 50, 100, 1000 times (once for each server)? 

Second, for the kinds of events you would get from OpenSea, it's a waste of resources. You've got an open socket sitting there doing nothing for most of the time. Each server instance has only so many available sockets and you need more load balancer bandwidth to handle all those constant connections.

Third, if a message is sent and is lost for some reason it's gone. 

The correct solution for this is URL callback registration where a client registers a URL with Open Sea to be called when an event happens.  The Open Sea servers would read from a queue of messages, attempt to send the message, if a 200 response is received delete the message and move on to the next. If no response or a failure response was received the message would go back to the queue for a retry later.  This way the Open Sea servers would need far less connections and the clients would not lose a message and the client's load balancer would route that message  only once to an available server.

**Comment by edgework** [comment]
> Why are websockets a bad idea?

As I said, this works fine if you have one server running in Heroku or something.  When you're a large company running many instances of the server in a scaling group you've got many websocket connections all getting the same message. How do you reconcile that? Do you update the local database with the same data 50, 100, 1000 times (once for each server)? 

Second, for the kinds of events you would get from OpenSea, it's a waste of resources. You've got an open socket sitting there doing nothing for most of the time. Each server instance has only so many available sockets and you need more load balancer bandwidth to handle all those constant connections.

Third, if a message is sent and is lost for some reason it's gone. 

The correct solution for this is URL callback registration where a client registers a URL with Open Sea to be called when an event happens.  The Open Sea servers would read from a queue of messages, attempt to send the message, if a 200 response is received delete the message and move on to the next. If no response or a failure response was received the message would go back to the queue for a retry later.  This way the Open Sea servers would need far less connections and the clients would not lose a message and the client's load balancer would route that message  only once to an available server.

**Comment by risingsun007** [comment]
The websocket seems to miss some messages

**Comment by rjbks** [comment]
@edgework why would you run multiple instances of this to receive and write the same events? Sounds like you're trying to scale a highly coupled monolithic app. Your solution, essentially asking OS to do some message routing on your behalf, wouldn't "solve" the problems you've mentioned anyway.

**Comment by risingsun007** [comment]
Why does the websocket miss events?

**Comment by edgework** [comment]
My solution is used by almost every major provider of anything. It's a tried and true architecture.  PayPal, Twilio, Stripe, Square, Amazon Storefront etc. etc. etc.   My solution is a whole lot easier for both Open Sea and its clients. They would be doing less work.   No large organization runs one of anything, there's redundancy, scaling and geolocation considerations.  And in my case, no, its not a "highly coupled monolithic app".

**Comment by rohfle** [comment]
If you have downtime and miss events via websockets, you can always grab event listings from the REST API for the downtime period and then process the ones you havent seen.

I've seen you talk about using webhooks, but these don't guarantee deliverability either, after some time the sender will delete messages it can't deliver.

Considering you are supposedly working at scale, if its such a big issue to you, why don't you roll your own websocket to webhook (or some other better solution for internal use) microservice and use that to guarantee deliverability, act as a proxy for your requests and deal with your downtime? You obviously have the time, money and knowledge to do this.

Not receiving events from the stream API is a separate problem, but this is not caused by the choice of websockets. I personally think its an internal issue with OpenSea's implementation but I am certain it can be fixed. Discussions are ongoing in #99

**Comment by edgework** [comment]
I'm not talking about long downtime, just those situations that come up now and then when the load balancer is reset or switched over; like 15 seconds maybe. Webhooks work great, most providers use them and they don't require a hack to get them to work.   And its pretty rare for messages to get dropped with a proper Webhook implementation.  I started up an old server once that interfaced Twilio and got messages from 5 months prior.  

...and you haven't addressed the issue of resource usage, and this is Open Sea's problem.  Lets see what happens when a bunch of creators (with varying levels of technical ability) are filling up their load balancer and servers with open sockets that just sit there.  I'm surprised whoever is running their dev ops didn't pitch a fit and refuse to allow it.  Unless their idea of dev ops is the last engineer to push to github.

**Comment by minerminer4949** [comment]
kafka? 
stream api -> kafka -> all your servers.

**Comment by edgework** [comment]
> kafka? stream api -> kafka -> all your servers.

Another workaround that requires more dev ops to maintain it.   And again, no organization runs one instance of anything so you're throwing multiple instancex of the same event into kafka and still need to deal with reconciling them.  I'm not even going to mention the hissy fit dev-ops will have when you want to throw a Java app at them.

**Comment by minerminer4949** [comment]
You're a fudder, plain and simple. Your comment about kafka reveals you have no idea how it works or how to scale apps. 
gn ser.

**Comment by edgework** [comment]
> your comment about kafka reveals you have no idea how it works or how to scale apps.

Maybe not but it's just yet another thing that needs to add complexity and management and cost  when simple WebHooks like everyone else  uses bypasses the need to even waste engineering resources to get up to speed and implement kafaka or whatever.  ...and when you're in an environment what does not use Java you're not going to get a Java app on it just to deal with a bad engineering decision, what's more likely (and I''m serious) is that last major deal is about to vanish because the CIO is heavily against it on technical grounds.

**Comment by Trontor** [comment]
> > your comment about kafka reveals you have no idea how it works
> 
> Maybe not but it's just yet another thing that needs to add complexity and management and cost when simple WebHooks like everyone else uses bypasses the need to even waste engineering resources to get up to speed and implement kafaka or whatever. ...and when you're in an environment what does not use Java you're not going to get a Java app on it just to deal with a bad engineering decision, what's more likely (and I''m serious) is that last major deal is about to vanish because the CIO is heavily against it on technical grounds.

How would webhooks solve your issue? Websockets are very performant and designed for multiplex high frequency delivery, what more can you ask for?

**Comment by edgework** [comment]
@Trontor Read the issue.

**Comment by Trontor** [comment]
> @Trontor Read the issue.

Your initial problem was "How do we reconcile the message when it's received 100 or 1000 times across 3 data centers? "

How does Webhooks solve this? You'll still receive the message across multiple data centers through the webhook?

**Comment by rick-rheo** [comment]
@Trontor   A load balancer would take the request and route it to only one server, like any other web request.

**Comment by Trontor** [comment]
@edgework so the data will enter a central point and it'll be redirected to the right server?

**Comment by edgework** [comment]
@Trontor It will be redirected to the next available server yeah.  

https://www.nginx.com/resources/glossary/round-robin-load-balancing/

**Comment by Trontor** [comment]
@edgework I'm aware what load balancing is. How exactly would you implement that for a Web hook?

**Comment by Trontor** [comment]
@edgework RFC 6455 states that WebSocket "is designed to work over HTTP ports 443 and 80 as well as to support HTTP proxies and intermediaries", thus making it compatible with HTTP.

**Comment by edgework** [comment]
A web socket is initiated from the client (in this case the server that needs events from OpenSea).  So yes a websocket will go though the Load Balancer (and keep a port constantly open which is wasting a resource) and each server instance will be creating a websocket since properly designed scalable servers have no state whatsoever and know absolutely nothing about any other instance running.  And there's the problem.

**Comment by edgework** [comment]
In a scaling environment I should be able to set up a load balancer and one or more servers behind it that have the same exact behavior and operate completely independently. The application (load balancer and servers) behavior should be exactly the same if there's one server or one thousand servers.

What you're talking about is some additional layer (application) to open fewer  Web Sockets and feed the events into some queue and reconcile them.  A large organization will NEVER run one instance of anything for redundancy, load spike cushion, etc. So with this solution we're going though the process of deploying a new application in the datacenter that needs to be managed. This adds significant costs and is a great reason for the head of Dev Ops to make noise and kill a deal with Open Sea when they understand why this application is necessary. Web Hooks are industry standard for sending events across services.  Web Sockets are for end-user interface clients.

**Comment by Trontor** [comment]
@edgework take Blocknative for example: https://www.blocknative.com/

this does the same thing. sends many events through a websocket. do you think that Blocknative is dumb as bricks for building their entire application on websockets? or perhaps, you're thinking about scalability incorrectly!

**Comment by edgework** [comment]
@Trontor Blocknative is a completely different use case.

**Comment by itsthedoom** [comment]
Hello frens, A few items to surface and highlight:
- The Stream API is in beta [Stream API Overview (Beta)](https://docs.opensea.io/reference/stream-api-overview#connecting--subscribing-to-events)
- There was a period when we were missing events. This has been resolved. [Missing events #99](https://github.com/ProjectOpenSea/stream-js/issues/99)

We're looking to dive deeper with devs and builders using the SDK and API; we would like to understand your use cases and implementation. Would you mind sharing additional details with us?

To help us out, please complete this form and include blockers/wish list items. https://airtable.com/shrvYIjP6Rt6p2jS6

**Comment by itsthedoom** [comment]
@edgework Thank you for the feedback. 

Would you mind summarizing the description in terms of action that we could take on this what is the biggest blocker you see?  

If multiple issues, could break them out into separate issues in order to help make this more actionable for us.

**Comment by edgework** [comment]
@itsthedoom   I think I stated the action and blockers in this issue more than once.

Implement Webhooks to send events to clients.  Its an industry standard for sending events across services.

**Comment by mar2194** [comment]
I agree that webhooks would be better for organizations that operate a scalable infrastructure and want more fine-grained control over resource management. The main issue here is that a websocket requires a 24/7 connection. Many orgs have a scalable container architecture for receiving webhooks from APIs that shutdown or spin up as demand ebbs and flows.

To do what devs looking for a webhook solution want to do with the OS stream API would require a dedicated container with an open websocket that then sends POST requests or pub/sub messages to other containers when a message is received. Essentially building our own OpenSea webhook API internally. 

While this is certainly feasible, it seems a bit unreasonable (and redundant) to expect multiple organizations to internally build the same product.

It seems like it would be somewhat trivial on the OpenSea side to setup a webhook API since the infrastructure is there for the websocket API.

Regarding a previous complaint about missing messages from the websocket API - no one should expect 100% delivery from webhooks or websockets. Websockets are no less reliable than webhooks and as someone else suggested, calling the HTTP API occasionally to verify completeness is probably a smart and reasonable approach for any organization receiving updates from webhooks or websockets.

Tldr: Please implement a Webhook API.

**Comment by sanguivore-easyco** [comment]
@itsthedoom a webhook API would be preferable for our needs as well, but barring that, we could more easily work around the issues with these changes:
- add unique ids to events (this would allow multiple listening-nodes to deduplicate data)
- add an HTTP endpoint that returns events with the same schema as the websocket (this would ease/allow backfills and patching up data that was sent during downtimes)
  - The current /events endpoint that I'm aware of has differing fields (e.g. "maker" and "taker" are "buyer" and "seller")

**Comment by edgework** [comment]
Still not solved.

**Comment by mar2194** [comment]
Are webhooks still not available from OpenSea? I gave up and just started using Moralis's and Alchemy's webhook APIs.
